### PR TITLE
Strict unmarshalling

### PR DESCRIPTION
**Pull request by jrudolph**
Only makes a difference with changes from https://github.com/akka/akka-http/pull/3822. Currently, akka-grpc never sees a strict entity.

### CODE REVIEW

File: runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala [hunk 1]
```diff
@@ -55,6 +61,12 @@ object GrpcMarshalling {
         .via(new CancellationBarrierGraphStage)
         .mapMaterializedValue(japiFunction(_ => NotUsed)))
   }
+  def unmarshalStream[T](
+      entity: HttpEntity,
+      u: ProtobufSerializer[T],
+      @silent("never used") mat: Materializer,
+      reader: GrpcProtocolReader): CompletionStage[Source[T, NotUsed]] =
```
Comment (viktorklang) [inline runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala:68]: @jrudolph `mat` is used.

Comment (jrudolph) [reply runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala:68]: Hah, copied the signature from above...

**Comment by johanandren** [comment]
Not sure if related to this or the marshalling optimization but with both PRs applied I get (both with nigthly snapshot of Akka HTTP and a snapshot with the improvements in https://github.com/akka/akka-http/pull/3822) this error when running the Akka Serverless TCK:

```
CANCELLED: Failed to read message.","exception":"io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at akka.grpc.internal.UnaryCallAdapter.onClose(UnaryCallAdapter.scala:40)
	at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:463)
	at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:427)
	at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:460)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: com.google.protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
	at com.google.protobuf.InvalidProtocolBufferException.truncatedMessage(InvalidProtocolBufferException.java:84)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.readRawBytes(CodedInputStream.java:1259)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.readBytes(CodedInputStream.java:913)
	at com.akkaserverless.protocol.discovery.Spec$.parseFrom(Spec.scala:126)
	at com.akkaserverless.protocol.discovery.Spec$.parseFrom(Spec.scala:113)
	at akka.grpc.scaladsl.ScalapbProtobufSerializer.deserialize(ScalapbProtobufSerializer.scala:19)
	at akka.grpc.scaladsl.ScalapbProtobufSerializer.deserialize(ScalapbProtobufSerializer.scala:14)
	at akka.grpc.internal.BaseMarshaller.parse(Marshallers.scala:48)\n\tat akka.grpc.internal.Marshaller.parse(Marshallers.scala:58)
	at akka.grpc.internal.Marshaller.parse(Marshallers.scala:56)\n\tat io.grpc.MethodDescriptor.parseResponse(MethodDescriptor.java:284)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInternal(ClientCallImpl.java:652)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:637)
	... 9 common frames omitted\n
```

No problems with only the Akka HTTP changes.

**Comment by viktorklang** [comment]
@johanandren Is this with the changes to BaseMarshaller, if so could you revert that change and see if that helps? Commit in question: https://github.com/akka/akka-grpc/commit/1aaea7f8b425e82966fde8b1a3051299fa3475c4

**Comment by raboof** [comment]
> @johanandren Is this with the changes to BaseMarshaller, if so could you revert that change and see if that helps? Commit in question: [1aaea7f](https://github.com/akka/akka-grpc/commit/1aaea7f8b425e82966fde8b1a3051299fa3475c4)

Hmm, good point, perhaps https://github.com/akka/akka-grpc/pull/1366/files#r644725655 ?

**Comment by johanandren** [comment]
That indeed seems to be the reason rather than these marshal/unmarshall changes. Thanks @viktorklang for figuring that out!

**Comment by viktorklang** [comment]
@johanandren @raboof No worries, I'm sorry for the breakage, and will curse at InputStream for a while this weekend.

**Comment by jrudolph** [comment]
All tests pass for me now (when you have access to the akka-http snapshot releases).

File: interop-tests/src/test/scala/akka/grpc/scaladsl/GrpcExceptionHandlerSpec.scala [hunk 2]
```diff
@@ -30,18 +30,22 @@ class GrpcExceptionHandlerSpec
     "produce an INVALID_ARGUMENT error when the expected parameter is not found" in {
       implicit val serializer = TestService.Serializers.SimpleRequestSerializer
       implicit val marshaller = GrpcProtocolNative.newWriter(Identity)
+      // request that missing the actual data
```
Comment (jrudolph) [inline interop-tests/src/test/scala/akka/grpc/scaladsl/GrpcExceptionHandlerSpec.scala:33]:
```suggestion
      // request that is missing the actual data
```

File: runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala [hunk 3]
```diff
@@ -55,6 +61,12 @@ object GrpcMarshalling {
         .via(new CancellationBarrierGraphStage)
         .mapMaterializedValue(japiFunction(_ => NotUsed)))
   }
+  def unmarshalStream[T](
+      entity: HttpEntity,
+      u: ProtobufSerializer[T],
+      @silent("never used") mat: Materializer,
```
Comment (jrudolph) [inline runtime/src/main/scala/akka/grpc/javadsl/GrpcMarshalling.scala:67]:
```suggestion
      mat: Materializer,
```

**Comment by jrudolph** [comment]
Ready from my side.

**Review by johanandren — APPROVED** [review]

**Pull request merged**